### PR TITLE
Fix issue 207

### PR DESCRIPTION
**Pull request by stephanenicolas**
Addresses #207 , #202, #236.

### CODE REVIEW

File: toothpick-compiler/src/main/java/toothpick/compiler/factory/FactoryProcessor.java [hunk 1]
```diff
@@ -83,6 +85,7 @@ public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment
 
     wasRun();
     readCommonProcessorOptions();
+    readCrashWhenNoFactoryCanBeCreatedOption();
```
Comment (dlemures) [inline toothpick-compiler/src/main/java/toothpick/compiler/factory/FactoryProcessor.java]: move inside `readCommonProcessorOptions()`

File: toothpick-compiler/src/main/java/toothpick/compiler/factory/FactoryProcessor.java [hunk 2]
```diff
@@ -297,14 +307,21 @@ private ConstructorInjectionTarget createConstructorInjectionTarget(TypeElement
       }
     }
 
+    final String cannotCreateAFactoryMessage = " Toothpick can't create a factory for it." //
+        + " If this class is itself a DI entry point (i.e. you call TP.inject(this) at some point), " //
+        + " then you can remove this warning by adding @SuppressWarnings(\"Injectable\") to the class." //
+        + " A typical example is a class using injection to assign its fields, that calls TP.inject(this)," //
```
Comment (dlemures) [inline toothpick-compiler/src/main/java/toothpick/compiler/factory/FactoryProcessor.java]: this second part looks to verbose

File: toothpick-compiler/src/main/java/toothpick/compiler/factory/FactoryProcessor.java [hunk 3]
```diff
@@ -397,4 +424,8 @@ void setToothpickRegistryChildrenPackageNameList(List<String> toothpickRegistryC
   void setToothpickExcludeFilters(String toothpickExcludeFilters) {
     this.toothpickExcludeFilters = toothpickExcludeFilters;
   }
+
+  public void setCrashWhenNoFactoryCanBeCreated(boolean crashWhenNoFactoryCanBeCreated) {
```
Comment (dlemures) [inline toothpick-compiler/src/main/java/toothpick/compiler/factory/FactoryProcessor.java]: can you make it package-protected? and add the `for testing only` comment

File: toothpick-compiler/src/main/java/toothpick/compiler/factory/FactoryProcessor.java [hunk 4]
```diff
@@ -297,14 +307,21 @@ private ConstructorInjectionTarget createConstructorInjectionTarget(TypeElement
       }
     }
 
+    final String cannotCreateAFactoryMessage = " Toothpick can't create a factory for it." //
```
Comment (dlemures) [inline toothpick-compiler/src/main/java/toothpick/compiler/factory/FactoryProcessor.java]: shouldn't this be a constant?

File: toothpick-compiler/src/main/java/toothpick/compiler/factory/FactoryProcessor.java [hunk 5]
```diff
@@ -297,14 +307,21 @@ private ConstructorInjectionTarget createConstructorInjectionTarget(TypeElement
...
-            warning(constructorElement, //
-                "The class %s has a private default constructor, Toothpick can't create a factory for it.",
-                typeElement.getQualifiedName().toString());
+              String message = format("The class %s has a private default constructor. "
+                      + cannotCreateAFactoryMessage, //
+                  typeElement.getQualifiedName().toString());
+            crashOrWarnWhenNoFactoryCanBeCreated(constructorElement, message);
           }
           return null;
         }
```
Comment (dlemures) [inline toothpick-compiler/src/main/java/toothpick/compiler/factory/FactoryProcessor.java]: we are repeating a bunch of code and looks complicated. Could we clean it a little bit on this PR?:
```java
for (ExecutableElement constructorElement : constructorElements) {
    if (constructorElement.getParameters().isEmpty()) {
        if(constructorElement.getModifiers().contains(Modifier.PRIVATE)) {
            break;
        }
        return new ConstructorInjectionTarget(typeElement, scopeName, hasScopeInstancesAnnotation, superClassWithInjectedMembers);
    }
}

if (!isInjectableWarningSuppressed(typeElement)) {
...

File: toothpick-compiler/src/main/java/toothpick/compiler/common/ToothpickProcessor.java [hunk 6]
```diff
@@ -76,13 +77,33 @@
...
+  /**
+   * The name of the annotation processor option to make the TP annotation processor crash when it detects
+   * an annotated method but with a non package-private visibility.
+   * By default the behavior is not to crash but emit a warning. Passing the value {@code true}
+   * crashes the build instead.
+   */
+  public static final String PARAMETER_CRASH_WHEN_INJECTED_METHOD_IS_NOT_PACKAGE = "toothpick_crash_when_injected_method_is_not_package";
+
+  /**  Allows to suppress warning when an injected method is not package-private visible. */
+  private static final String SUPPRESS_WARNING_ANNOTATION_VISIBLE_VALUE = "visible";
```
Comment (dlemures) [inline toothpick-compiler/src/main/java/toothpick/compiler/common/ToothpickProcessor.java]: move to the memberinjector processor plz

File: toothpick-compiler/src/main/java/toothpick/compiler/common/ToothpickProcessor.java [hunk 7]
```diff
@@ -147,6 +168,7 @@ protected void readCommonProcessorOptions() {
     readOptionRegistryPackageName();
     readOptionRegistryChildrenPackageNames();
     readOptionExcludes();
+    readOptionCrashWhenMethodIsNotPackageProtected();
```
Comment (dlemures) [inline toothpick-compiler/src/main/java/toothpick/compiler/common/ToothpickProcessor.java]: same, just for mi

File: toothpick-compiler/src/main/java/toothpick/compiler/common/ToothpickProcessor.java [hunk 8]
```diff
@@ -193,20 +215,35 @@ protected void readOptionAnnotationTypes() {
     }
   }
 
+  private void readOptionCrashWhenMethodIsNotPackageProtected() {
```
Comment (dlemures) [inline toothpick-compiler/src/main/java/toothpick/compiler/common/ToothpickProcessor.java]: same

File: toothpick-compiler/src/main/java/toothpick/compiler/common/ToothpickProcessor.java [hunk 9]
```diff
@@ -76,13 +77,33 @@
...
+   */
+  public static final String PARAMETER_CRASH_WHEN_NO_FACTORY_CAN_BE_CREATED = "toothpick_crash_when_no_factory_can_be_created";
+
+  /**
+   * The name of the annotation processor option to make the TP annotation processor crash when it detects
+   * an annotated method but with a non package-private visibility.
+   * By default the behavior is not to crash but emit a warning. Passing the value {@code true}
+   * crashes the build instead.
+   */
+  public static final String PARAMETER_CRASH_WHEN_INJECTED_METHOD_IS_NOT_PACKAGE = "toothpick_crash_when_injected_method_is_not_package";
```
Comment (dlemures) [inline toothpick-compiler/src/main/java/toothpick/compiler/common/ToothpickProcessor.java]: this is not set in the MI processor as a supported option

File: toothpick-compiler/src/main/java/toothpick/compiler/common/ToothpickProcessor.java [hunk 10]
```diff
@@ -193,20 +215,35 @@ protected void readOptionAnnotationTypes() {
     }
   }
 
+  private void readOptionCrashWhenMethodIsNotPackageProtected() {
+    Map<String, String> options = processingEnv.getOptions();
+    if(toothpickCrashWhenMethodIsNotPackageVisible == null) {
+      toothpickCrashWhenMethodIsNotPackageVisible = Boolean.parseBoolean(options.get(PARAMETER_CRASH_WHEN_INJECTED_METHOD_IS_NOT_PACKAGE));
+    }
+  }
```
Comment (dlemures) [inline toothpick-compiler/src/main/java/toothpick/compiler/common/ToothpickProcessor.java]: we are doing it for both processor, but we only listen to the annotation on the MI?

File: toothpick-compiler/src/main/java/toothpick/compiler/memberinjector/MemberInjectorProcessor.java [hunk 11]
```diff
@@ -187,4 +187,8 @@ void setToothpickRegistryChildrenPackageNameList(List<String> toothpickRegistryC
   void setToothpickExcludeFilters(String toothpickExcludeFilters) {
     this.toothpickExcludeFilters = toothpickExcludeFilters;
   }
+
+  void setCrashOrWarnWhenMethodIsNotPackageVisible(boolean crashOrWarnWhenMethodIsNotPackageVisible) {
```
Comment (dlemures) [inline toothpick-compiler/src/main/java/toothpick/compiler/memberinjector/MemberInjectorProcessor.java]: add testing only comment

File: toothpick-compiler/src/main/java/toothpick/compiler/common/ToothpickProcessor.java (hunk 9 above)
Comment (stephanenicolas) [inline toothpick-compiler/src/main/java/toothpick/compiler/common/ToothpickProcessor.java]: done

**Pull request merged**